src/components/FallenHeroes.tsx

```diff
@@ -11,32 +11,46 @@ function FallenHeroes() {
   // matching hero from the set of totalHeroes,
   // and is included in rest of heroes alive, update the state accordingly
   const { campaign, setCampaign } = React.useContext(LotrContext);
+
+  function filterFallen() {
+    return campaign.allHeroes.filter((hero) => hero.alive === false);
+  }
+
+  const fallen = filterFallen();
   const [fallenHero, setFallenHero] = useState("");
 
+  const names = campaign.allHeroes.map((hero) => {
+    return hero.name;
+  });
+
+  const isAlive =
+    campaign.allHeroes.find((hero) => hero.name === fallenHero)?.alive === true;
+
   function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
     setFallenHero(e.target.value);
   }
 
   function addFallenHero(e: React.FormEvent<HTMLFormElement>) {
     e.preventDefault();
     //add hero to the list if it isnt there yet
-    if (!campaign.restOfAliveHeroes.includes(fallenHero)) {
+    if (!names.includes(fallenHero)) {
+      throw Error("Sorry, your hero is an unknown!");
+    }
+
+    if (names.includes(fallenHero) && !isAlive) {
       throw Error("Sorry, your hero is not among the living!");
     }
-    if (campaign.restOfAliveHeroes.includes(fallenHero)) {
-      const updatedFallen = [...campaign.fallenHeroes, fallenHero];
-      const updatedRestOfAlive = campaign.restOfAliveHeroes.filter(
-        (hero) => hero !== fallenHero
-      );
-      setCampaign({
-        ...campaign,
-        fallenHeroes: updatedFallen,
-        restOfAliveHeroes: updatedRestOfAlive,
+
+    if (names.includes(fallenHero) && isAlive) {
+      campaign.allHeroes.find((hero) => {
```

Comment:
See Eslint complaining here? The `find` callback expects a return value but this one doesn't have any. This way, the loop doesn't stop when the fallen hero is found. Instead, assign the found hero to a variable and use it later.

Notice that you are searching for the same hero on line 27. If you use an intermediary variable inside the `addFallenHero` function, one search is enough.

Reply:
nothing complaining here as well. But maybe this later version is ok..